torch/csrc/jit/init.cpp

```diff
@@ -406,23 +407,65 @@ void initJITBindings(PyObject* module) {
 
   py::class_<PyTorchStreamWriter>(m, "PyTorchFileWriter")
       .def(py::init<std::string>())
-      .def(
-          "write_record",
-          [](PyTorchStreamWriter& self,
-             const std::string& name,
-             const char* data,
-             size_t size) { return self.writeRecord(name, data, size); })
-      .def("write_end_of_file", &PyTorchStreamWriter::writeEndOfFile);
+      .def(py::init([](const py::object &buffer) {
+        auto writer_func = [=](const void *data, size_t size) {
+          auto bytes = py::bytes(reinterpret_cast<const char *>(data), size);
+          buffer.attr("write")(std::move(bytes));
+          return size;
+        };
+        return caffe2::make_unique<PyTorchStreamWriter>(std::move(writer_func));
+      }))
+      .def(py::init<const std::function<size_t(const void *, size_t)> &>())
+      .def("write_record",
+           [](PyTorchStreamWriter &self, const std::string &name,
+              const char *data,
+              size_t size) { return self.writeRecord(name, data, size); })
+      .def("write_end_of_file", &PyTorchStreamWriter::writeEndOfFile)
+      .def("write_record",
+           [](PyTorchStreamWriter &self, const std::string &name,
+              uintptr_t data, size_t size) {
+             return self.writeRecord(name, reinterpret_cast<const char *>(data),
+                                     size);
+           });
 
   py::class_<PyTorchStreamReader>(m, "PyTorchFileReader")
       .def(py::init<std::string>())
-      .def("get_record", [](PyTorchStreamReader& self, const std::string& key) {
+      .def(py::init([](const py::object &buffer) {
+        // Get size
```

Comment:
Is it required to know size for the reader protocol? This means you can't use this reader for true streams, where seeking to the end would result in materializing all the data.

Reply:
(I guess the protocol here requires random access seeking, so this is fine)

Reply:
Miniz requires knowing the size of whatever its data buffer is (I guess so it can jump to the central directory of the zip file